Mathlib/CategoryTheory/LocallyCartesianClosed/ExponentiableMorphism.lean

```diff
@@ -0,0 +1,180 @@
+  ev f |>.naturality g
+
+theorem coev_naturality {X Y : Over J} (g : X ⟶ Y) :
+    g ≫ (coev f).app Y = (coev f).app X ≫ (pushforward f).map ((pullback f).map g) :=
+  coev f |>.naturality g
+
+/-- The first triangle identity for the counit and unit of the adjunction. -/
+@[reassoc]
+theorem ev_coev (X : Over J) :
+    (pullback f).map (coev f |>.app X) ≫ (ev f |>.app (pullback f |>.obj X)) =
+    𝟙 (pullback f |>.obj X) :=
+  pullbackAdjPushforward f |>.left_triangle_components X
+
+/-- The second triangle identity for the counit and unit of the adjunction. -/
+@[reassoc]
+theorem coev_ev (Y : Over I) :
+    (coev f |>.app (pushforward f |>.obj Y)) ≫
+    (pushforward f |>.map (ev f |>.app Y)) =
+    𝟙 (pushforward f |>.obj Y) :=
+  pullbackAdjPushforward f |>.right_triangle_components Y
+
+variable {f}
+
+/-- The currying of `(pullback f).obj A ⟶ X` in `Over I` to a morphism `A ⟶ (pushforward f).obj X`
+in `Over J`. -/
+def pushforwardCurry {X : Over I} {A : Over J}
+    (u : (pullback f).obj A ⟶ X) :
+    A ⟶ (pushforward f).obj X :=
+  pullbackAdjPushforward f |>.homEquiv A X u
+
+/-- The uncurrying of `A ⟶ (pushforward f).obj X` in `Over J` to a morphism
+`(Over.pullback f).obj A ⟶ X` in `Over I`. -/
+def pushforwardUncurry {X : Over I} {A : Over J}
+    (v : A ⟶ (pushforward f).obj X) :
+    (pullback f).obj A ⟶ X :=
+  pullbackAdjPushforward f |>.homEquiv A X |>.invFun v
+
+theorem homEquiv_apply_eq {X : Over I} {A : Over J} (u : (pullback f).obj A ⟶ X) :
+    (pullbackAdjPushforward f |>.homEquiv _ _) u = pushforwardCurry u :=
+  rfl
+
+theorem homEquiv_symm_apply_eq {X : Over I} {A : Over J} (v : A ⟶ (pushforward f).obj X) :
+    (pullbackAdjPushforward f |>.homEquiv _ _).symm v = pushforwardUncurry v :=
+  rfl
+
+theorem pushforward_uncurry_curry {X : Over I} {A : Over J}
+    (u : (pullback f).obj A ⟶ X) :
+    pushforwardUncurry (pushforwardCurry u) = u :=
+  pullbackAdjPushforward f |>.homEquiv A X |>.left_inv u
+
+theorem pushforward_curry_uncurry {X : Over I} {A : Over J} (v : A ⟶ (pushforward f).obj X) :
+    pushforwardCurry (pushforwardUncurry v) = v :=
+  pullbackAdjPushforward f |>.homEquiv A X |>.right_inv v
+
+instance : ChosenPullbacksAlong (Over.mk f).hom := by
+  dsimp only [Over.mk_hom]
+  infer_instance
+
+instance OverMkHom : ExponentiableMorphism (Over.mk f).hom := by
+  dsimp only [Over.mk_hom]
+  infer_instance
+
+end
+
+section
+
+/-- The identity morphisms `𝟙 _` are exponentiable. -/
+@[simps]
+def id (I : C) [ChosenPullbacksAlong (𝟙 I)] : ExponentiableMorphism (𝟙 I) :=
+  ⟨𝟭 _, ofNatIsoLeft (F := 𝟭 _) Adjunction.id (pullbackId I).symm⟩
+
+/-- Any pushforward of the identity morphism is naturally isomorphic to the identity functor. -/
+def pushforwardId (I : C) [ChosenPullbacksAlong (𝟙 I)] [ExponentiableMorphism (𝟙 I)] :
+    pushforward (𝟙 I) ≅ 𝟭 (Over I) :=
+  Adjunction.rightAdjointUniq (pullbackAdjPushforward (𝟙 I)) (id I).pullbackAdjPushforward
+
```

Comment:
I think it might be useful to have the same simp lemmas as for `ChosenPullbacks` that characterizes this isomorphism (e.g `unit_pushforwardId_hom_app` etc.), and same for `comp`